python/tvm/relay/frontend/caffe.py

```diff
@@ -0,0 +1,1188 @@
+
+    for pl in predict_layer:
+        if len(pl.bottom) > 0:
+            for t in pl.top:
+                if t not in not_outputs:
+                    model_outputs.append(t)
+    return model_inputs, model_outputs
+
+
+def from_caffe(init_net, predict_net, shape_dict, dtype_dict):
+    """Convert from caffe model into compatible relay Function.
+
+    Parameters
+    ----------
+    init_net : caffe_pb2.NetParameter
+        caffemodel
+    predict_net : caffe_pb2.NetParameter
+        caffe prototxt
+    shape_dict : dict of str to int list/tuple
+        Input shapes of the model.
+    dtype_dict : dict of str to str
+        Input types of the model.
+
+    Returns
+    -------
+    mod : tvm.relay.Module
+        The relay module for compilation.
+
+    params : dict of str to tvm.NDArray
+        The parameter dict to be used by relay
+    """
+    logging.debug('caffe frontend')
+
+    old_caffe = False
+    if len(predict_net.input) != 0:  # old caffe version
+        old_caffe = True
+        model_inputs = list(predict_net.input)
+
+    predict_layer = predict_net.layer
+
+    # replace layer's top with its name and update other layers'bottoms
+    top_change_before_dict = _rebuild_layers(predict_layer)
+
+    # obtain inputs and outputs of Net
+    if old_caffe:
+        _, model_outputs = _get_inputs_outputs(predict_layer)
+    else:
+        model_inputs, model_outputs = _get_inputs_outputs(predict_layer)
+
+    logging.debug('model_inputs:%s', ",".join(model_inputs))
+    logging.debug('model_outputs:%s', ','.join(model_outputs))
+
+    exp_tab = ExprTable()
+    for in_name in model_inputs:
+        shape = shape_dict[in_name] if in_name in shape_dict else None
+        dtype = dtype_dict[in_name] if in_name in dtype_dict else "float32"
+        exp_tab.set_expr(in_name, _expr.var(in_name, shape=shape, dtype=dtype))
+
+    init_layer = init_net.layer
+    init_layer_dict = {il.name: il for il in init_layer}
+    predict_layer_dict = {pl.name: pl for pl in predict_layer}
+    # op code in model
+    op_converter = OperatorConverter(init_layer_dict, predict_layer, exp_tab)
+    op_converter.check_unsupported_ops()
+    op_converter.op_fuse()
+    op_converter.convert_op_to_relay()
+
+    # params and outputs
+    params = {k: _nd.array(np.array(v)) for k, v in exp_tab.params.items()}
+    outputs = list()
+    for n in model_outputs:
+        if n in op_converter.changed_layers:
+            n = op_converter.changed_layers[n]
+        outputs.append(exp_tab.get_expr(n))
+    outputs = outputs[0] if len(outputs) == 1 else _expr.Tuple(outputs)
+    func = _function.Function(analysis.free_vars(outputs), outputs)
+
+    mod = IRModule.from_expr(func)
+    # return mod, params
+    new_outputs = list()
+    for i in model_outputs:
+        if i in top_change_before_dict:
+            tmp = top_change_before_dict[i]
+        else:
+            tmp = i
+        new_outputs.append(tmp)
+    return mod, params, new_outputs
```

Comment:
Please refer out other frontend code and we should only return `mod` and `params`.